datafusion-postgres/src/handlers.rs

```diff
@@ -83,8 +83,27 @@ impl SimpleQueryHandler for DfSessionService {
             .await
             .map_err(|e| PgWireError::ApiError(Box::new(e)))?;
 
-        let resp = datatypes::encode_dataframe(df, &Format::UnifiedText).await?;
-        Ok(vec![Response::Query(resp)])
+        let query_lower = query.to_lowercase();
+        if query_lower.starts_with("insert into") {
+            // For INSERT queries, we need to execute the query to get the row count
+            // and return an Execution response with the proper tag
+            let result = df
+                .clone()
+                .collect()
+                .await
+                .map_err(|e| PgWireError::ApiError(Box::new(e)))?;
+
+            // Get the number of rows affected (typically 1 for INSERT)
+            let rows_affected = result.iter().map(|batch| batch.num_rows()).sum::<usize>();
+
+            // Create INSERT tag with the affected row count
+            let tag = Tag::new("INSERT").with_oid(0).with_rows(rows_affected);
+            Ok(vec![Response::Execution(tag)])
+        } else {
+            // For non-INSERT queries, return a regular Query response
+            let resp = datatypes::encode_dataframe(df, &Format::UnifiedText).await?;
+            Ok(vec![Response::Query(resp)])
+        }
```

Comment:
I think this patch returns total count of rows stored in current datafusion context. However that's different semantics from return value of `INSERT` statements. What if the database already has a few rows?

Also I think datafusion is a read-only query engine so it won't process insert statement. If current error message will cause some issue with postgres client, we can check the query before running `ctx.sql(query)` and skip it.

Reply:
OK. I just realized datafusion is no longer read-only and can do create table and insert statements.

Here you will need to use the return value of INSERT statement and wrap it into postgres tag. See the return value:

```
public=> INSERT INTO delhi_2013 SELECT date, meantemp, humidity, meanpressure FROM delhi WHERE date < '2014-05-07';
 count
-------
   491
(1 row)
```

Reply:
counter example: count of 1 even though 3 records inserted
![CleanShot 2025-04-14 at 14 56 09@2x](https://github.com/user-attachments/assets/ce5c1de7-fd54-4d9e-9645-3870b51e5476)